wordpress-mobile/WordPress-iOS
# Extract web authentication logic

This is the first step towards a simpler, more modern, web view controller. In this PR, I've extracted all the authentication logic to a separate `WebViewAuthenticator` class that has unit tests.

Some decisions that I made that might need explanation:

We still can't authenticate Jetpack sites, but we're treating them as WordPress.com for authentication purposes, because that at least enables SSO to work. It's not the ideal experience, but it's a one tap login, and then the site will be authenticated in subsequent views until the cookies expire.

In some places that are not site related (stats, reader, notifications) we're pre-authenticating with the account. The rationale is that any of those links could be a WordPress.com site even without a `.wordpress.com` domain. This means that we're also doing login when someone visits a link to an external site (e.g. github.com). WordPress.com login won't allow an open redirect, so I had to come up with a hack: it'll redirect to `https://wordpress.com/?wpios_redirect=<the_actual_redirect_to>`, and we intercept that in the web view delegate. Not 100% happy with the solution, but it was either that, or adding a lot of logic to figure out if something was a WordPress.com URL (not 100% accurate).

Improves: #7761 as the theme detail is logged in, but activating doesn't take you back to the app.
Improves: #7778 #7759 as you still get the login form, but if there's a SSO you are already logged in to WordPress.com
Related: #7781

To test:

Verify that all instances of WPWebViewController work as expected. Some examples:

- Click a link on the reader, comment, stats
- View Site
- Set up a new sharing connection for a site

To help with testing/debugging, if you want to clear cookies, you can pause the app, and use this in the debugger console:

```
(lldb) po [[[NSHTTPCookieStorage sharedHTTPCookieStorage] cookies] count]
6

(lldb) po [[NSHTTPCookieStorage sharedHTTPCookieStorage] removeCookiesSinceDate:[NSDate dateWithTimeIntervalSinceReferenceDate:0]]
(lldb) po [[[NSHTTPCookieStorage sharedHTTPCookieStorage] cookies] count]
<nil>
```


Needs review: @elibud
